src/arrow/schema.rs

```diff
@@ -0,0 +1,992 @@
+      }
+    } else {
+      false
+    }
+  }
+  
+  fn is_repeated(&self) -> bool {
+    let basic_info = self.schema.get_basic_info();
+    
+    basic_info.has_repetition() && basic_info.repetition()==Repetition::REPEATED
+  }
+  
+  fn is_self_included(&self) -> bool {
+    self.leaves.contains(&(self.schema.as_ref() as *const Type))
+  }
+  
+  
+  fn to_group_type(&self) -> Result<Option<DataType>> {
+    if self.is_repeated() {
+      self.to_struct().map(|opt| opt.map(|dt| DataType::List(Box::new(dt))))
+    } else {
+      match self.schema.get_basic_info().logical_type() {
+        LogicalType::LIST => self.to_list(),
+        _ => self.to_struct()
+      }
+    }
+  }
+  
+  // Functions for primitive types.
+  
+  fn to_primitive_type(&self) -> Result<Option<DataType>> {
+    if self.is_self_included() {
+      self.to_primitive_type_inner().map(|dt| {
+        if self.is_repeated() {
+          Some(DataType::List(Box::new(dt)))
+        } else {
+          Some(dt)
+        }
+      })
+    } else {
+      Ok(None)
+    }
+  }
+  
+  fn to_primitive_type_inner(&self) -> Result<DataType> {
+    match self.schema.get_physical_type() {
+      PhysicalType::BOOLEAN => Ok(DataType::Boolean),
+      PhysicalType::INT32 => self.to_int32(),
+      PhysicalType::INT64 => self.to_int64(),
+      PhysicalType::FLOAT => Ok(DataType::Float32),
+      PhysicalType::DOUBLE => Ok(DataType::Float64),
+      PhysicalType::BYTE_ARRAY => self.to_byte_array(),
+      other => Err(ArrowError(format!("Unable to convert parquet type {}", other)))
+    }
+  }
+  
+  fn to_int32(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::NONE => Ok(DataType::Int32),
+      LogicalType::UINT_8 => Ok(DataType::UInt8),
+      LogicalType::UINT_16 => Ok(DataType::UInt16),
+      LogicalType::UINT_32 => Ok(DataType::UInt32),
+      LogicalType::INT_8 => Ok(DataType::Int8),
+      LogicalType::INT_16 => Ok(DataType::Int16),
+      LogicalType::INT_32 => Ok(DataType::Int32),
+      other => Err(ArrowError(
+        format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  fn to_int64(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::NONE => Ok(DataType::Int64),
+      LogicalType::INT_64 => Ok(DataType::Int64),
+      LogicalType::UINT_64 => Ok(DataType::UInt64),
+      other => Err(ArrowError(
+        format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  fn to_byte_array(&self) -> Result<DataType> {
+    match self.schema.get_basic_info().logical_type() {
+      LogicalType::UTF8 => Ok(DataType::Utf8),
+      other => Err(ArrowError(
+        format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  // Functions for group types.
+  
+  fn to_struct(&self) -> Result<Option<DataType>> {
+    match self.schema.as_ref() {
+      Type::PrimitiveType {..} => panic!("This should not happen."),
```

Comment:
Please, extend these error messages, they should indicate the error and potentially include metadata that could aid debugging.